kore/src/Kore/Step/Simplification/NoConfusion.hs

```diff
@@ -70,22 +139,16 @@ to be different; therefore their conjunction is @\\bottom@.
 -}
 constructorAndEqualsAssumesDifferentHeads ::
     MonadUnify unifier =>
-    HasCallStack =>
     TermLike RewritingVariableName ->
     TermLike RewritingVariableName ->
-    MaybeT unifier a
+    unifier a
 constructorAndEqualsAssumesDifferentHeads
-    first@(App_ firstHead _)
-    second@(App_ secondHead _) =
+    first
+    second =
         do
-            Monad.guard =<< Simplifier.isConstructorOrOverloaded firstHead
-            Monad.guard =<< Simplifier.isConstructorOrOverloaded secondHead
-            assert (firstHead /= secondHead) $
-                lift $ do
-                    explainBottom
-                        "Cannot unify different constructors or incompatible \
-                        \sort injections."
-                        first
-                        second
-                    empty
-constructorAndEqualsAssumesDifferentHeads _ _ = empty
+            explainBottom
+                "Cannot unify different constructors or incompatible \
+                \sort injections."
+                first
+                second
```

Comment:
I don't remember, what does `explainBottom` do exactly? Does it throw an error or something or do we handle whatever it throws somewhere?

Reply:
This is addressed by https://github.com/kframework/kore/issues/2639